stdlib/public/core/CollectionDifference.swift

```diff
@@ -0,0 +1,353 @@
+  ///   between two states.
+  ///
+  /// - Complexity: O(*n* * log(*n*)), where *n* is the length of the
+  ///   parameter.
+  init<C>(validatedChanges c: C) where C:Collection, C.Element == Change {
+    let changes = c.sorted { (a, b) -> Bool in
+      switch (a, b) {
+      case (.remove(_, _, _), .insert(_, _, _)):
+        return true
+      case (.insert(_, _, _), .remove(_, _, _)):
+        return false
+      default:
+        return a.offset < b.offset
+      }
+    }
+
+    // Find first insertion via binary search
+    let firstInsertIndex: Int
+    if changes.count == 0 {
+      firstInsertIndex = 0
+    } else {
+      var range = 0...changes.count
+      while range.lowerBound != range.upperBound {
+        let i = (range.lowerBound + range.upperBound) / 2
+        switch changes[i] {
+        case .insert(_, _, _):
+          range = range.lowerBound...i
+        case .remove(_, _, _):
+          range = (i + 1)...range.upperBound
+        }
+      }
+      firstInsertIndex = range.lowerBound
+    }
+
+    removals = Array(changes[0..<firstInsertIndex])
+    insertions = Array(changes[firstInsertIndex..<changes.count])
+  }
+
+  /// The `.insert` changes contained by this difference, from lowest offset to highest
+  public let insertions: [Change]
+  
+  /// The `.remove` changes contained by this difference, from lowest offset to highest
+  public let removals: [Change]
+}
+
+/// A CollectionDifference is itself a Collection.
+///
+/// The enumeration order of `Change` elements is:
+///
+/// 1. `.remove`s, from highest `offset` to lowest
+/// 2. `.insert`s, from lowest `offset` to highest
+///
+/// This guarantees that applicators on compatible base states are safe when
+/// written in the form:
+///
+/// ```
+/// for c in diff {
+///   switch c {
+///   case .remove(offset: let o, element: _, associatedWith: _):
+///     arr.remove(at: o)
+///   case .insert(offset: let o, element: let e, associatedWith: _):
+///     arr.insert(e, at: o)
+///   }
+/// }
+/// ```
+extension CollectionDifference : Collection {
+  public typealias Element = CollectionDifference<ChangeElement>.Change
+
+  // Opaque index type is isomorphic to Int
+  public struct Index: Comparable, Hashable {
+    public static func < (lhs: CollectionDifference<ChangeElement>.Index, rhs: CollectionDifference<ChangeElement>.Index) -> Bool {
+      return lhs.i < rhs.i
+    }
+    
+    let i: Int
+    init(_ index: Int) {
+      i = index
+    }
+  }
+
+  public var startIndex: CollectionDifference<ChangeElement>.Index {
+    return Index(0)
+  }
+  
+  public var endIndex: CollectionDifference<ChangeElement>.Index {
+    return Index(removals.count + insertions.count)
+  }
+  
+  public func index(after index: CollectionDifference<ChangeElement>.Index) -> CollectionDifference<ChangeElement>.Index {
+    return Index(index.i + 1)
+  }
+  
+  public subscript(position: CollectionDifference<ChangeElement>.Index) -> Element {
```

Comment:
I don't see the line width in the browser, but you should get an idea by now of what should be fixed.